### ISSUE TITLE
custom `numpy` dtypes

### ISSUE THREAD

**Issue by keewis** [geoxarray/geoxarray#20]
I wonder if it would be worth investigating if it is possible to implement a custom `numpy` `dtype` (see NEP 40-43), similar to `geopandas`' `GeometryArray`? We can probably try to interface with `odc-geo` for this.

There's not a lot of documentation so far (besides the NEPs), but there *is* a [experimentation repository](https://github.com/seberg/unitdtype) that can serve as a reference. It is written in C, though.

**Comment by snowman2** [comment]
I wonder if this is relevant: https://github.com/pydata/xarray/pull/5102

**Comment by djhoese** [comment]
Yeah I was thinking the same thing @snowman2. I think because we aren't dealing with dataframes, we'd need to talk about coordinates or indexes of a certain numpy dtype.

**Comment by keewis** [comment]
I can't point you to a specific place, but it seems there's work in progress to use `awkward` to represent polylines / polygons in memory which would make this suggestion obsolete.

Regarding the index: what exactly did you have in mind? There is [a temporary API](https://github.com/pydata/xarray/blob/scipy22/xarray/core/dataset.py#L4147-L4223) to set indexes, so we could start working on that (and the more indexes we have, the easier it will be to define the final public API).

**Comment by keewis** [comment]
see pydata/xarray#4285 for some related discussion

**Comment by keewis** [comment]
`shapely` 2.0 apparently will be exactly that: `numpy` arrays of geometries with operations as `numpy` ufuncs (though those arrays will still have the `object` dtype).

See also [this discussion on the pangeo forum](https://discourse.pangeo.io/t/vector-data-cubes/2904/13).
